crates/anvil-polkadot/src/substrate_node/service/mod.rs

```diff
@@ -48,11 +69,228 @@ pub struct Service {
+    let keys = fetch_all_keys_paged(client, at_hex).await?;
+
+    let pb = ProgressBar::new(keys.len() as u64);
+    pb.set_style(
+        ProgressStyle::with_template("[{elapsed_precise}] {bar:40.cyan/blue} {pos}/{len} values")
+            .unwrap()
+            .progress_chars("=>-"),
+    );
+    pb.set_message("Downloading values...");
+
+    let mut top_map: Map<String, Value> = Map::new();
+    for k in keys.iter() {
+        let v: Option<String> =
+            client.request("state_getStorage", rpc_params![k.clone(), at_hex]).await?;
+        if let Some(val_hex) = v {
+            top_map.insert(k.clone(), Value::String(val_hex));
+        }
+        pb.inc(1);
+    }
+
+    pb.finish_with_message("All values downloaded ✔");
+    Ok(top_map)
+}
+
+fn build_forked_chainspec_from_raw_top(
+    top_map: Map<String, Value>,
+) -> sc_service::error::Result<Box<dyn sc_chain_spec::ChainSpec>> {
+    let children_default = serde_json::Map::<String, Value>::new();
+
+    let spec_json = json!({
+        "name": "Anvil Polkadot (Forked)",
+        "id": "anvil-polkadot-forked",
+        "chainType": "Development",
+        "bootNodes": [],
+        "telemetryEndpoints": null,
+        "protocolId": null,
+        "properties": null,
+        "codeSubstitutes": {},
+        "consensusEngine": null,
+        "genesis": { "raw": { "top": top_map, "childrenDefault": children_default }}
+    });
+
+    let bytes = serde_json::to_vec(&spec_json)
+        .map_err(|e| ServiceError::Other(format!("serialize spec json failed: {e}")))?;
+    type EmptyExt = Option<()>;
+    let new_spec: sc_chain_spec::GenericChainSpec<EmptyExt> =
+        sc_chain_spec::GenericChainSpec::from_json_bytes(bytes)
+            .map_err(|e| ServiceError::Other(format!("from_json_bytes failed: {e}")))?;
+    Ok(Box::new(new_spec))
+}
+
+fn create_manual_seal_inherent_data_providers(
+    client: Arc<Client>,
+    time_manager: Arc<TimeManager>,
+) -> impl Fn(
+    Hash,
+    (),
+) -> futures::future::Ready<
+    Result<
+        (sp_timestamp::InherentDataProvider, MockValidationDataInherentDataProvider<()>),
+        Box<dyn std::error::Error + Send + Sync>,
+    >,
+> + Send
++ Sync {
+    move |block: Hash, ()| {
+        let current_para_head = client
+            .header(block)
+            .expect("Header lookup should succeed")
+            .expect("Header passed in as parent should be present in backend.");
+
+        let slot_duration = client.runtime_api().slot_duration(current_para_head.hash()).unwrap();
+
+        let para_id = client.runtime_api().parachain_id(current_para_head.hash()).unwrap();
+
+        let current_para_block_head =
+            Some(polkadot_primitives::HeadData(current_para_head.encode()));
+
+        let next_block_number =
+            UniqueSaturatedInto::<u32>::unique_saturated_into(current_para_head.number) + 1;
+
+        let next_time = time_manager.current_call_timestamp().checked_mul(1000).unwrap();
```

Comment:
We'll need to do: 
```suggestion
        let next_time = time_manager.next_timestamp().checked_mul(1000).unwrap();
```
The way we do it now will not advance time.